### PR TITLE
Close #21728: Apply "Fix all rides" even if mechanic is already working

### LINKED ISSUE #21728: "Fix all rides" cheat ignores rides currently being fixed by mechanic
### Operating System

Ubuntu Linux 22.04.4

### OpenRCT2 build

develop 318bff1, Linux AppImage

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

_No response_

### Describe the issue

Broken down rides that are currently being fixed by a mechanic will _not_ be fixed by clicking the "Fix all rides" cheat.

Attached park has a ride being fixed, which will ignore clicking the cheat's button.

### Steps to reproduce

- Wait for or force a ride breakdown
- Wait for mechanic to start fixing the ride
- Click "Fix all rides" cheat
- Ride stays broken until the mechanic finishes fixing the ride

### Attachments

[fix_all_rides_cheat_ignores_rides_being_fixed.park.zip](https://github.com/OpenRCT2/OpenRCT2/files/14884010/fix_all_rides_cheat_ignores_rides_being_fixed.park.zip)

### PR DESCRIPTION
**Pull request by JDCustard**
Fixes #21728 Fix all rides cheat ignores when ride is being fixed my mechanic

### CODE REVIEW

**Comment by Gymnasiast** [comment]
Next time, please don’t close a PR just to open a new one. It causes noise and loses all the conversation.

**Comment by JDCustard** [comment]
> Next time, please don’t close a PR just to open a new one. It causes noise and loses all the conversation.

Noted. Apologies.

File: src/openrct2/actions/CheatSetAction.cpp [hunk 1]
```diff
@@ -512,13 +512,23 @@ void CheatSetAction::FixBrokenRides() const
...
+
+            if (ride.mechanic_status == RIDE_MECHANIC_STATUS_FIXING)
             {
-                mechanic->RemoveFromRide();
+                if (mechanic != nullptr)
+                {
+                    mechanic->RideSubState = PeepRideSubState::ApproachExitWaypoints;
+                }
+            }
+            else
```
Comment (Gymnasiast) [inline src/openrct2/actions/CheatSetAction.cpp:526]: We’ll probably want to limit this `else` to "responding to breakdown call" and "answering breakdown call". It prevents mechanics heading for an inspection from being unceremoniously kicked from the job. (I know that that’s what it’s already doing on develop, but this seemed like a good opportunity to change it.)

File: src/openrct2/actions/CheatSetAction.cpp [hunk 2]
```diff
@@ -512,13 +512,23 @@ void CheatSetAction::FixBrokenRides() const
...
             auto mechanic = RideGetAssignedMechanic(ride);
-            if (mechanic != nullptr)
+
+            if (ride.mechanic_status == RIDE_MECHANIC_STATUS_FIXING)
             {
-                mechanic->RemoveFromRide();
+                if (mechanic != nullptr)
+                {
+                    mechanic->RideSubState = PeepRideSubState::ApproachExitWaypoints;
+                }
```
Comment (ZehMatt) [inline src/openrct2/actions/CheatSetAction.cpp:524]: Its probably best to check if mechanic is nullptr after line 517 and continue the loop if it is null.

File: src/openrct2/actions/CheatSetAction.cpp [hunk 3]
```diff
@@ -512,13 +512,23 @@ void CheatSetAction::FixBrokenRides() const
...
+            }
+            else
+            {
+                if (mechanic != nullptr)
+                {
+                    mechanic->RemoveFromRide();
+                }
```
Comment (ZehMatt) [inline src/openrct2/actions/CheatSetAction.cpp:531]: See my comment above, no need to have the null check in two branches.

**Review by ZehMatt — CHANGES_REQUESTED** [review]

File: src/openrct2/actions/CheatSetAction.cpp (hunk 1 above)
Comment (JDCustard) [reply src/openrct2/actions/CheatSetAction.cpp:526]: > We’ll probably want to limit this `else` to "responding to breakdown call" and "answering breakdown call". It prevents mechanics heading for an inspection from being unceremoniously kicked from the job. (I know that that’s what it’s already doing on develop, but this seemed like a good opportunity to change it.)

Yeah that is a good point, we don't want mechanics inspecting or on the way to inspect to be removed from the ride. Thank you.

**Review by ZehMatt — APPROVED** [review]
Code looks alright, needs a bit of testing I think.

**Comment by cheweytoo** [comment]
I've done some testing with the github AppImage build: Looks good.

- The cheat now instantly fixes rides that already have a mechanic fixing them (tested with up to three simultaneous repairs going on).
- The fixing mechanic gets properly interrupted and leaves the ride

Thank you!

**Review by tupaschoal — APPROVED** [review]

**Pull request merged**